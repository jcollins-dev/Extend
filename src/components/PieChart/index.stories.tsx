import React from 'react';
import { Meta } from '@storybook/react/types-6-0';
import PieChart, { PieSliceDatum } from '.';

const testData: PieSliceDatum[] = [
  {
    color: 'hsl(222, 70%, 50%)',
    hidden: false,
    id: '801',
    label: '4.2 (20.9%)',
    percent: 20.9,
    tooltip: <span>20.9%</span>
  },
  {
    color: 'hsl(84, 70%, 50%)',
    hidden: false,
    id: '230',
    label: '2.7 (13.5%)',
    percent: 13.5,
    tooltip: <span>13.5%</span>
  },
  {
    color: 'hsl(307, 70%, 50%)',
    hidden: false,
    id: '1',
    label: '2.7 (13.4%)',
    percent: 13.4,
    tooltip: <span>13.4%</span>
  },
  {
    color: 'hsl(169, 70%, 50%)',
    hidden: false,
    id: '101',
    label: '2.4 (11.7%)',
    percent: 11.7,
    tooltip: <span>11.7%</span>
  },

  {
    color: '#000000',
    hidden: false,
    id: 'other',
    label: '8.2 (40.6%)',
    percent: 40.6,
    tooltip: <span>40.6%</span>
  }
];

export default {
  title: 'Charting/PieChart',
  component: PieChart
} as Meta;

export const SimplePieChart = (): JSX.Element => (
  <div style={{ width: '400px' }}>
    <PieChart data={testData} />
  </div>
);
