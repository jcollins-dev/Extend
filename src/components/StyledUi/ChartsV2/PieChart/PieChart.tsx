//3rd party
import React from 'react';
import { VictoryPie } from 'victory';
import { DarkTooltip, DefaultTooltip, LightTooltip } from '../TooltipsV1';

/** data being sent to pie chart in the required format */
/** data must be converted before sending to PieChart */
export interface PieChartDataProps {
  x: string;
  y: number;
  label?: string;
}

/** click handers for pie slicers, returns the X: label value of the pie slice. */
export type PieChartHandlerProps = (x: string) => void;
/** array of labels that are selected, can be sent from the filters array from useFilterStates, it's used for styling */
export type PieChartSelectedProps = string[];
/** colors that are assigned to each slice/by idKey (x: in the pie chart array of objects) */
export type PieChartColorsProps = { [key: string]: string };
/** is this a doughnut style pie chart?  defaults to standard pie chart */
export type PieChartIsDoughnutProps = boolean;

export type LegendProps = {
  name?: string;
  symbol?: {
    fill?: string;
    type?: string;
  };
};

export interface PieChartProps {
  /** incoming aray of data objects in the form of [{x: label, y: value}] */
  data: PieChartDataProps[];
  handleClick?: PieChartHandlerProps;
  selected?: PieChartSelectedProps;
  colors?: PieChartColorsProps;
  isDoughnut?: PieChartIsDoughnutProps;
  height?: number;
  width?: number;
  padding?: number;
  standAlone?: boolean;
  labelComponent?: JSX.Element;
  hasLegend?: boolean;
  tooltipType?: `light` | `dark`;
}

export const PieChart = ({
  data,
  handleClick,
  selected,
  colors,
  isDoughnut,
  tooltipType,
  height,
  width,
  padding,
  standAlone,
  labelComponent
}: PieChartProps): JSX.Element => {
  // return nothing if no data
  if (!data) return <></>;
  let BarTooltip = DefaultTooltip;
  if (tooltipType === 'dark') BarTooltip = DarkTooltip;
  if (tooltipType === 'light') BarTooltip = LightTooltip;

  return (
    <VictoryPie
      standalone={standAlone}
      data={data}
      width={width}
      height={height || width}
      padding={padding}
      labelRadius={120}
      labelComponent={labelComponent || BarTooltip}
      labels={({ datum }) => `x: ${datum.x}, y: ${datum.y}`}
      innerRadius={(/*{ datum }*/) => (isDoughnut ? 50 : 0)}
      colorScale={!colors ? 'qualitative' : undefined}
      style={{
        data: {
          // if filters are active and if label is on filter list, change the opacity of the slice to indicate
          // it's not selected
          cursor: 'pointer',
          transition: 'fill-opacity 300ms ease',
          fillOpacity: ({ datum }) =>
            !selected?.length ? 1 : selected?.includes(datum.x) ? 1 : 0.2
        }
      }}
      events={[
        {
          childName: 'all',
          target: 'data',
          eventHandlers: {
            onClick: () => ({
              // get the x value and send to click handlers as string
              mutation: ({ datum }: { datum: PieChartDataProps }) =>
                handleClick?.(datum.x as string)
            }) /*,
            
            //// will need this later ////

            onMouseOver: (_, obj) => {
              console.log(obj)
              return {

                // get the x value and send to click handlers as string
                mutation: (over) =>
                  console.log({ over })
              }
            },
            onMouseOut: () => ({}),
            
            /////////////////////////////////

            */
          }
        }
      ]}
    />
  );
};
