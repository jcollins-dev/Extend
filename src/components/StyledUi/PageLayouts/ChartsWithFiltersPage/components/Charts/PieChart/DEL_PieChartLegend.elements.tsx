import styled from 'styled-components';

export const PieChartLegendContainer = styled.ul`
  li {
    list-style: none;
    padding: 0.3em 0;
    margin: 0;
    display: grid;
    grid-template-columns: auto 1fr;
    gap: 0.7em;
    font-size: 1.1em;

    &[data-muted='true'] {
      opacity: 0.5;
    }

    button {
      display: flex;
      gap: 0.5em;
      border: none;
      background-color: transparent;
      font-family: inherit;
      cursor: pointer;
      text-align: left;
      align-items: center;

      &:before {
        content: '';
        width: 1em;
        height: 1em;
        display: block;
        border-radius: 4px;
      }
    }

    .item-count {
      opacity: 0.8;
      font-size: 0.9em;
      display: flex;

      &:before {
        content: '(';
      }

      &:after {
        content: ')';
      }
    }
  }
`;
interface ColorProps {
  color?: string;
}
export const PieChartLegendItemContainer = styled.li<ColorProps>`
  button:before {
    background-color: ${({ color }) => color};
  }
`;
