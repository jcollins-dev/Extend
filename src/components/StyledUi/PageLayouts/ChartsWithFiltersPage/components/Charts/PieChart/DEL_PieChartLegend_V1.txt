import React from 'react';
import { PieChartLegendContainer, PieChartLegendItemContainer } from './PieChartLegend.elements';

export interface PieChartLegendProps {
  legendItems: {
    label: string;
    color?: string;
  }[];
  legendItemsSelected?: string[];
  handleLegendClick: (label: string) => void;
}
export const PieChartLegend = ({
  legendItems,
  legendItemsSelected,
  handleLegendClick
}: PieChartLegendProps): JSX.Element => {
  return (
    <PieChartLegendContainer className="pie-chart-legend">
      {legendItems?.map(({ label, color }) => {
        const isSelected = !legendItemsSelected ? true : legendItemsSelected?.includes(label);
        const useLabel =
          label === 'Critical'
            ? 'A-Alarm (Critical)'
            : label === 'Product'
            ? 'B-Alarm (Product)'
            : label === 'Warning'
            ? 'C-Alarm (Warning)'
            : label;
        return (
          <PieChartLegendItemContainer
            key={label}
            color={color}
            className="pie-chart-legend__item"
            data-muted={!isSelected ? 'true' : undefined}
          >
            <button type="button" onClick={() => handleLegendClick?.(label)}>
              {useLabel} {/*showCount && y && <span className="item-count">{y}</span>*/}
            </button>
          </PieChartLegendItemContainer>
        );
      })}
    </PieChartLegendContainer>
  );
};
