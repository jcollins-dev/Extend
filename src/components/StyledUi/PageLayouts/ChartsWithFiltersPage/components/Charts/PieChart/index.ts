export * from './PieChart';
export * from './PieChartTooltips';
