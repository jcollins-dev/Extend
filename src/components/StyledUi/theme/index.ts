import * as base from './standard.theme';
import { font } from './fonts';
import light from './light.theme';
import * as effects from './effects';
import * as charts from './charts';

export const styledTheme = { effects, color: light, font, ...charts, ...base };
