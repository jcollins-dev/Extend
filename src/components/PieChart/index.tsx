// 3rd party libs
import React, { useState } from 'react';
import { VictoryPie, VictoryLabel, VictoryContainer } from 'victory';

// Components
import { ChartToolTip } from 'components';

// Types
import { ChartToolTipProps } from 'components/ChartToolTip';

// Theme
import theme from 'themes';

export interface PieSliceDatum {
  id: string;
  label?: string;
  percent: number;
  color: string;
  tooltip?: React.ReactNode;
  hidden?: boolean;
  // Highlight slice
  highlight?: boolean;
}

interface Props {
  data: PieSliceDatum[];
  onSliceClick?: (datum: PieSliceDatum) => void;
}

const RADIUS = 90;
const PADDING = { top: 10, left: 30, right: 30, bottom: 30 };
const MIN_SIZE_SHOW_LABEL = 12;

const PieChart = ({ data, onSliceClick }: Props): JSX.Element => {
  // Ref to chart container allows us to find its position on the page (using getBoundingClientRect)
  const chartRef = React.useRef<HTMLDivElement>(null);

  const [toolTipState, setToolTipState] = useState<ChartToolTipProps | null>(null);

  /**
   * Shows a tooltip for a pie slice.
   * The posiiton is calculated by the provided angles of the slice.
   */
  const showTooltip = (startAngle: number, endAngle: number, content: React.ReactNode) => {
    // How far from the centre should the tooltip locate
    const toolTipPositionRadius = RADIUS * 0.5;

    // The arc of the slice that has been selected, in degrees
    const arcDegs = endAngle - startAngle;

    // Angle at the center of the arc, in degrees
    const angleDegs = startAngle + arcDegs / 2;

    // Convert to radians
    const angleRads = (angleDegs / 180) * Math.PI;

    // Calculate x/y position of the tooltip (with respect to the chart container)
    const x = toolTipPositionRadius * Math.sin(angleRads);
    const y = toolTipPositionRadius * Math.cos(angleRads);

    // Get the position of the chart on the page
    const chartPosition = chartRef.current?.getBoundingClientRect();

    if (chartPosition) {
      // Set the absolute tooltip position on the page
      setToolTipState({
        x: chartPosition.left + window.scrollX + PADDING.left + RADIUS + x,
        y: chartPosition.top + window.scrollY + PADDING.top + RADIUS - y,
        content
      });
    }
  };

  const transformedData = () =>
    data.map((a) => ({ ...a, label: `${parseFloat(a.label ?? '').toFixed(1)}%` }));

  return (
    <>
      <div ref={chartRef}>
        <VictoryPie
          containerComponent={<VictoryContainer responsive={false} />}
          y="percent"
          labelRadius={RADIUS * 0.65}
          width={RADIUS * 2 + PADDING.left + PADDING.right}
          height={RADIUS * 2 + PADDING.top + PADDING.bottom}
          padding={PADDING}
          data={transformedData()}
          labelComponent={
            <VictoryLabel
              backgroundComponent={<rect ry={5} rx={5} />}
              backgroundPadding={5}
              backgroundStyle={{
                fill: theme.colors.black,
                opacity: ({ datum }) =>
                  datum.hidden || datum.percent < MIN_SIZE_SHOW_LABEL ? 0 : 0.5,
                pointerEvents: 'none'
              }}
              style={{
                fill: theme.colors.white,
                fontSize: '10',
                pointerEvents: 'none',
                opacity: ({ datum }) =>
                  datum.hidden || datum.percent < MIN_SIZE_SHOW_LABEL ? 0 : 1
              }}
            />
          }
          events={[
            {
              target: 'data',
              eventHandlers: {
                onMouseOver: (_, obj) => {
                  showTooltip(obj.datum.startAngle, obj.datum.endAngle, obj.datum.tooltip);
                },
                onMouseOut: () => {
                  setToolTipState(null);
                },
                onClick: (_, obj) => {
                  onSliceClick && onSliceClick(obj.datum);
                }
              }
            }
          ]}
          style={{
            data: {
              fill: ({ datum }) => (datum.hidden ? 'transparent' : datum.color),
              cursor: 'pointer',
              stroke: ({ datum }) => datum.highlight && datum.color,
              strokeWidth: ({ datum }) => datum.highlight && 5,
              pointerEvents: ({ datum }) => (datum.hidden ? 'none' : 'auto')
            }
          }}
        />
      </div>
      {toolTipState && toolTipState.content && <ChartToolTip {...toolTipState} />}
    </>
  );
};

export default PieChart;
