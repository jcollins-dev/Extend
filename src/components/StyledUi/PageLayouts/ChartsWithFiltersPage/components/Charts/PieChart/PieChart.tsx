import React, { useState } from 'react';
import { VictoryPie, VictoryLabel } from 'victory';

import { ChartTooltipWrapper, ChartTooltipWrapperPropsTooltip } from '../../../components';
import { GlobalChartProps } from '../Charts.types';
import { globalChartStyles, globalChartHandlers } from '../_helpers';

export interface PieChartProps extends GlobalChartProps {
  data?: Record<string, unknown>[];
  barRatio?: number;
  labelRadius?: number;
}

export const PieChart = ({
  data,
  dimensions,
  colors,
  handleClick,
  checkIfSelected,
  colorScale,
  colorKey,
  CustomTooltip
}: PieChartProps): JSX.Element => {
  const [tooltipProps, setTooltipProps] = useState<ChartTooltipWrapperPropsTooltip | undefined>(
    undefined
  );

  if (tooltipProps) {
    console.log('tooltip not set', { setTooltipProps });
  }
  // return empty if no data.
  // this should be checked for higher up, in case data is loading.
  // if this loads and there is no data, there should a message saying no data
  // in a parent component

  if (!data) return <>error no data</>;

  const width = dimensions?.width;
  const height = dimensions?.height;

  const handleHover = () => {
    /*const slice = props?.slice as Record<string, number>;
    const datum = props?.datum as Record<string, unknown>;

    const calculatePieSliceCenter = (startAngle: number, endAngle: number) => {
      const chartWidth = 175;
      const chartHeight = 175;
      const containerLeft = 0;
      const containerTop = 0;

      const chartCenterX = containerLeft + chartWidth / 2;
      const chartCenterY = containerTop + chartHeight / 2;
      const sliceAngle = (startAngle + endAngle) / 2;

      const radius = Math.min(chartWidth, chartHeight) / 2;
      const sliceCenterX = chartCenterX + radius * Math.cos(sliceAngle * (Math.PI / 180));
      const sliceCenterY = chartCenterY + radius * Math.sin(sliceAngle * (Math.PI / 180));

      return { x: sliceCenterX, y: sliceCenterY };
    };*/

    //if (slice && width && height && datum) {
    //const center = calculatePieSliceCenter(slice.startAngle, slice.endAngle); //({ x: width / 2 + (width / 2) * Math.cos((slice.startAngle + slice.endAngle) / 2 * (Math.PI / 180)), y: height / 2 + (height / 2) * Math.sin((slice.startAngle + slice.endAngle) / 2 * (Math.PI / 180)) });      console.log('center not set', { center });
    //return setTooltipProps({ x: center.x, y: center.y, datum })
    //}
    return;
  };

  const CustomLabel = (props: Record<string, unknown>) => {
    const datum = props.datum as Record<string, unknown>;
    //const y = datum.y as number;

    const settings: Record<string, unknown> = {
      renderInPortal: true,
      textAnchor: 'middle', //"middle" | "inherit" | "start" | "end"
      verticleAnchor: 'middle', //| "start" | "end"
      x: props.x as number,
      y: props.y as number,
      text: `${datum.y}%`,
      backgroundStyle: {
        fill: 'black',
        filter: 'drop-shadow(0px 3px 2px rgba(0, 0, 0, .3))',
        rx: 5,
        opacity:
          (datum.y as number) < 10 ? 0 : checkIfSelected?.({ label: datum?.label }) ? 0.4 : 0.1
      },
      backgroundPadding: { left: 5, top: 3, right: 5, bottom: 3 },

      style: {
        fontFamily: 'inherit',
        fill: 'white',
        fontWeight: 500,
        opacity: (datum.y as number) < 10 ? 0 : checkIfSelected?.({ label: datum?.label }) ? 1 : 0.3
      }
    };

    /* this removes the background style of smaller values,
    /* please keep for now, will be needed in the near future
    // thanks TS
    if (y < 15) {
      delete settings.backgroundStyle;
      delete settings.backgroundPadding;
      delete settings.backgroundComponent;
    }
    */
    return <VictoryLabel {...settings} />;
  };

  return (
    <>
      {CustomTooltip && <ChartTooltipWrapper {...{ tooltipProps, CustomTooltip, handleHover }} />}

      <VictoryPie
        {...{
          width,
          height,
          data,
          style: globalChartStyles({ checkIfSelected, colors, colorScale, colorKey }),
          events: globalChartHandlers({ handleClick, handleHover }),
          domainPadding: 0,
          padding: 0,
          labelComponent: <CustomLabel />,
          //labelComponent: <CustomLabel />,
          labelRadius: Number(width) * 0.3
        }}
      />
    </>
  );
};
