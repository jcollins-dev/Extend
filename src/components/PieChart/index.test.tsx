import React from 'react';
import { render } from '@testing-library/react';
import { ThemeProvider } from 'styled-components';
import theme from 'themes';
import PieChart, { PieSliceDatum } from '.';

const testData: PieSliceDatum[] = [
  {
    id: 'slice1',
    label: '35%',
    percent: 35,
    color: '#ff0000',
    tooltip: <span>slice1: 35%</span>
  },
  {
    id: 'slice2',
    label: '35%',
    percent: 35,
    color: '#ff0000',
    tooltip: <span>slice2: 35%</span>
  },
  {
    id: 'slice3',
    label: '30%',
    percent: 30,
    color: '#ff0000',
    tooltip: <span>slice3: 30%</span>
  }
];

describe('PieChart', () => {
  it('Matches snapshot when rendering with data', () => {
    const { container } = render(
      <ThemeProvider theme={theme}>
        <PieChart data={testData} />
      </ThemeProvider>
    );

    expect(container).toMatchSnapshot();
  });
});
