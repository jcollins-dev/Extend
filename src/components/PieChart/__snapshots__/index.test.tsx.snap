// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PieChart Matches snapshot when rendering with data 1`] = `
<div>
  <div>
    <div
      class="VictoryContainer"
      style="pointer-events: none; position: relative; width: 240px; height: 220px;"
    >
      <svg
        height="220"
        role="img"
        style="pointer-events: all; width: 240px; height: 220px;"
        width="240"
      >
        <g>
          <path
            d="M5.5109105961630896e-15,-90A90,90,0,0,1,72.81152949374527,52.90067270632258L0,0Z"
            role="presentation"
            shape-rendering="auto"
            style="fill: #ff0000; cursor: pointer; pointer-events: auto; padding: 10px;"
            transform="translate(120, 100)"
          />
          <path
            d="M72.81152949374527,52.90067270632258A90,90,0,0,1,-85.59508646656381,27.811529493745276L0,0Z"
            role="presentation"
            shape-rendering="auto"
            style="fill: #ff0000; cursor: pointer; pointer-events: auto; padding: 10px;"
            transform="translate(120, 100)"
          />
          <path
            d="M-85.59508646656381,27.811529493745276A90,90,0,0,1,-1.6532731788489267e-14,-90L0,0Z"
            role="presentation"
            shape-rendering="auto"
            style="fill: #ff0000; cursor: pointer; pointer-events: auto; padding: 10px;"
            transform="translate(120, 100)"
          />
          <g>
            <rect
              height="17.788375000000002"
              rx="5"
              ry="5"
              style="fill: #000000; opacity: 0.5; pointer-events: none;"
              width="33.96875"
              x="167"
              y="69"
            />
            <text
              direction="inherit"
              dx="0"
              id="pie-labels-0"
              x="172"
              y="76.55"
            >
              <tspan
                dx="0"
                dy="5"
                style="fill: #FFFFFF; font-size: 10px; pointer-events: none; opacity: 1; font-family: 'Gill Sans', 'Gill Sans MT', 'Ser­avek', 'Trebuchet MS', sans-serif; stroke: transparent;"
                text-anchor="start"
                x="172"
              >
                35.0%
              </tspan>
            </text>
          </g>
          <g>
            <rect
              height="17.788375000000002"
              rx="5"
              ry="5"
              style="fill: #000000; opacity: 0.5; pointer-events: none;"
              width="33.96875"
              x="94"
              y="158"
            />
            <text
              direction="inherit"
              dx="0"
              id="pie-labels-1"
              x="111"
              y="166.55"
            >
              <tspan
                dx="0"
                dy="5"
                style="fill: #FFFFFF; font-size: 10px; pointer-events: none; opacity: 1; font-family: 'Gill Sans', 'Gill Sans MT', 'Ser­avek', 'Trebuchet MS', sans-serif; stroke: transparent;"
                text-anchor="middle"
                x="111"
              >
                35.0%
              </tspan>
            </text>
          </g>
          <g>
            <rect
              height="17.788375000000002"
              rx="5"
              ry="5"
              style="fill: #000000; opacity: 0.5; pointer-events: none;"
              width="33.96875"
              x="44"
              y="62"
            />
            <text
              direction="inherit"
              dx="0"
              id="pie-labels-2"
              x="73"
              y="69.55"
            >
              <tspan
                dx="0"
                dy="5"
                style="fill: #FFFFFF; font-size: 10px; pointer-events: none; opacity: 1; font-family: 'Gill Sans', 'Gill Sans MT', 'Ser­avek', 'Trebuchet MS', sans-serif; stroke: transparent;"
                text-anchor="end"
                x="73"
              >
                30.0%
              </tspan>
            </text>
          </g>
        </g>
      </svg>
      <div
        style="z-index: 99; position: absolute; top: 0px; left: 0px; width: 240px; height: 220px;"
      >
        <svg
          height="220"
          style="overflow: visible; width: 240px; height: 220px;"
          width="240"
        />
      </div>
    </div>
  </div>
</div>
`;
